orchestrator/services/subscription_relations.py

```diff
@@ -0,0 +1,269 @@
+    subscription_depends_on_ids: dict[UUID, list[UUID]] = {}
+    for relation in depends_on_relations:
+        subscription_depends_on_ids.setdefault(relation.in_use_by_sub_id, []).append(relation.depends_on_sub_id)
+
+    def get_depends_on_subs(in_use_by_id: UUID) -> list[SubscriptionTable]:
+        depends_on_ids = subscription_depends_on_ids.get(in_use_by_id, [])
+        return [depends_on_sub for id_ in depends_on_ids if (depends_on_sub := depends_on_subs.get(id_))]
+
+    # Important (as with any dataloader)
+    # Return the list of dependson subs in the exact same order as the ids passed to this function
+    return [get_depends_on_subs(subscription_id) for subscription_id in subscription_ids]
+
+
+async def get_recursive_relations(
+    subscription_ids: list[UUID],
+    filter_statuses: tuple[str, ...],
+    recurse_product_types: list[str],
+    recurse_depth_limit: int,
+    relation_fetcher: Callable[[list[UUID], tuple[str, ...]], Awaitable[list[list[SubscriptionTable]]]],
+    current_depth: int = 0,
+    used_subscription_ids: set[UUID] | None = None,
+) -> list[SubscriptionTable]:
+    """Recursively fetches subscription relations based on a custom relation-fetching function.
+
+    Args:
+        subscription_ids: List of subscription IDs to start with.
+        filter_statuses: List of statuses to filter the relations.
+        recurse_product_types: List of product types to recurse into.
+        recurse_depth_limit: Maximum recursion depth.
+        current_depth: Current recursion depth.
+        used_subscription_ids: List of already used subscription IDs to avoid cycles.
+        relation_fetcher: A function that fetches the relations for the subscriptions.
+
+    Returns:
+        A flattened list of all fetched relations for the given subscriptions.
+    """
+
+    used_subscription_ids = used_subscription_ids or set(subscription_ids)
+    relations = await relation_fetcher(subscription_ids, filter_statuses)
+    flat_relations_list = list(flatten(relations))
+
+    def get_related_subscription_ids() -> list[UUID]:
+        return [
+            r.subscription_id
+            for r in flat_relations_list
+            if r.product.product_type in recurse_product_types and r.subscription_id not in used_subscription_ids
+        ]
+
+    nested_relations: list[SubscriptionTable] = []
+    if recurse_depth_limit > current_depth and (related_subscription_ids := get_related_subscription_ids()):
+        used_subscription_ids.update(related_subscription_ids)
+        nested_relations = await get_recursive_relations(
+            related_subscription_ids,
+            filter_statuses,
+            recurse_product_types,
+            recurse_depth_limit,
+            current_depth=current_depth + 1,
+            used_subscription_ids=used_subscription_ids,
+            relation_fetcher=relation_fetcher,
+        )
+    return list(unique(flat_relations_list + nested_relations, lambda s: s.subscription_id))
```

Comment:
Some suggestions:
* Use `unique_everseen` instead of `unique` here to save time on sorting
* Use `itertools.chain` to avoid creating a new list
* Pass sort-key parameter as a keyword arg
```suggestion
    return list(unique_everseen(chain(relations, nested_relations), key=lambda s: s.subscription_id))
```